pkg/assembler/backends/ent/backend/hasMetadata.go

```diff
@@ -0,0 +1,262 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backend
+
+import (
+	"context"
+	stdsql "database/sql"
+
+	"entgo.io/ent/dialect/sql"
+	"github.com/guacsec/guac/pkg/assembler/backends/ent"
+	"github.com/guacsec/guac/pkg/assembler/backends/ent/hasmetadata"
+	"github.com/guacsec/guac/pkg/assembler/backends/ent/predicate"
+	"github.com/guacsec/guac/pkg/assembler/backends/helper"
+	"github.com/guacsec/guac/pkg/assembler/graphql/model"
+	"github.com/pkg/errors"
+	"github.com/vektah/gqlparser/v2/gqlerror"
+)
+
+func (b *EntBackend) HasMetadata(ctx context.Context, filter *model.HasMetadataSpec) ([]*model.HasMetadata, error) {
+	records, err := b.client.HasMetadata.Query().
+		Where(hasMetadataPredicate(filter)).
+		Limit(MaxPageSize).
+		WithSource(withSourceNameTreeQuery()).
+		WithArtifact().
+		WithPackageVersion(withPackageVersionTree()).
+		WithAllVersions(withPackageNameTree()).
+		All(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return collect(records, toModelHasMetadata), nil
+}
+
+func (b *EntBackend) IngestHasMetadata(ctx context.Context, subject model.PackageSourceOrArtifactInput, pkgMatchType *model.MatchFlags, hasMetadata model.HasMetadataInputSpec) (*model.HasMetadata, error) {
+	recordID, err := WithinTX(ctx, b.client, func(ctx context.Context) (*int, error) {
+		return upsertHasMetadata(ctx, ent.TxFromContext(ctx), subject, pkgMatchType, hasMetadata)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &model.HasMetadata{ID: nodeID(*recordID)}, nil
+}
+
+func (b *EntBackend) IngestBulkHasMetadata(ctx context.Context, subjects model.PackageSourceOrArtifactInputs, pkgMatchType *model.MatchFlags, hasMetadataList []*model.HasMetadataInputSpec) ([]string, error) {
+	var results []string
+	for i := range hasMetadataList {
```

Comment:
qq: is there a more efficient way to do bulk ingestion in ENT or is looping through and calling `IngestHasMetadata` the only way?

Reply:
That's a good question.
It's not the more efficient way and I've already investigated on how to improve it.
I'm just waiting for all the Ent endpoints to be available (with this "basic" iterative bulk approach) and then we'll move into optimizing them, starting from a consistent situation.

Basically we're trying to reach completeness before moving into further optimizations in order to avoid losing control of the situation.